File: code/py/ch11/11.01.py
## 11.1 Kerasによる回帰

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from keras import activations, callbacks, layers, models
from sklearn.preprocessing import StandardScaler
from sklearn.utils import shuffle

my_url = ('https://raw.githubusercontent.com/taroyabuki'
          '/fromzero/master/data/wine.csv')
tmp = pd.read_csv(my_url)

my_data = shuffle(tmp)

my_scaler = StandardScaler()
X = my_scaler.fit_transform(
    my_data.drop(columns=['LPRICE2']))
y = my_data['LPRICE2']

x = np.linspace(-3, 3, 100)
plt.plot(x, activations.relu(x))
plt.xlabel('x')
plt.ylabel('ReLU(x)')

my_model = models.Sequential()
my_model.add(layers.Dense(units=3, activation='relu', input_shape=[4]))
my_model.add(layers.Dense(units=1))

my_model.summary() # ネットワークの概要
#> Model: "sequential"
#> _________________________________________________________________
#> Layer (type)                 Output Shape              Param #
#> =================================================================
#> dense (Dense)                (None, 3)                 15
#> _________________________________________________________________
#> dense_1 (Dense)              (None, 1)                 4
#> =================================================================
#> Total params: 19
#> Trainable params: 19
#> Non-trainable params: 0

my_model.compile(
    loss='mse',
    optimizer='rmsprop')

my_cb = callbacks.EarlyStopping(
    patience=20,
    restore_best_weights=True)

my_history = my_model.fit(
    x=X,
    y=y,
    validation_split=0.25,
    batch_size=10,
    epochs=500,
    callbacks=[my_cb],
    verbose=0)

tmp = pd.DataFrame(my_history.history)
tmp.plot(xlabel='epoch')

tmp.iloc[-1, ]
#> loss        0.192743
#> val_loss    0.342249
#> Name: 499, dtype: float64

y_ = my_model.predict(X)
((y_.ravel() - y)**2).mean()
#> 0.23050613964540986

File: code/R/ch11/11.01.R
## 11.1 Kerasによる回帰

library(keras)
library(tidyverse)
my_url <- str_c("https://raw.githubusercontent.com/taroyabuki",
                "/fromzero/master/data/wine.csv")
tmp <- read_csv(my_url)

my_data <- tmp[sample(nrow(tmp)), ]

X <- my_data %>%
  select(-LPRICE2) %>% scale
y <- my_data$LPRICE2

curve(activation_relu(x), -3, 3)

my_model <- keras_model_sequential() %>%
  layer_dense(units = 3, activation = "relu", input_shape = c(4)) %>%
  layer_dense(units = 1)

summary(my_model) # ネットワークの概要
# 割愛（Pythonの結果を参照）

my_model %>% compile(
  loss = "mse",
  optimizer = "rmsprop")

my_cb <- callback_early_stopping(
  patience = 20,
  restore_best_weights = TRUE)

my_history <- my_model %>% fit(
    x = X,
    y = y,
    validation_split = 0.25,
    batch_size = 10,
    epochs = 500,
    callbacks = list(my_cb),
    verbose = 0)

plot(my_history)

my_history
#> Final epoch (plot to see history):
#>     loss: 0.06124
#> val_loss: 0.1132

y_ <- my_model %>% predict(X)
mean((y_ - y)^2)**0.5
#> [1] 0.2724372
